runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler.cpp

```diff
@@ -216,6 +224,56 @@ OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_initCifNativeTh
 	ffiTypeHelpers.freeStructFFIType(returnType);
 	goto done;
 }
+
+#if JAVA_SPEC_VERSION >= 22
+/* openj9.internal.foreign.abi.InternalDowncallHandler: private static native boolean isFfiProtoEnabled(); */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_isFfiProtoEnabled(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	jboolean isFfiProtoOn = J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_FFI_PROTO) ? JNI_TRUE : JNI_FALSE;
+
+	VM_OutOfLineINL_Helpers::returnSingle(currentThread, isFfiProtoOn, 0);
+	return rc;
+}
+
+/**
+ * openj9.internal.foreign.abi.InternalDowncallHandler: private static native void setNativeInvokeCache(MethodHandle handle);
+ *
+ * @brief Resolve the bounded handle via MethodHandleResolver.ffiCallLinkCallerMethod() in advance
+ *        when calling into the interpreter via MH.linkToNative()
+ *
+ * @return void
+ */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_setNativeInvokeCache(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	j9object_t invokeCache = NULL;
+	j9object_t nativeMH = J9_JNI_UNWRAP_REFERENCE(currentThread->sp);
+
+	/* The resolution is performed by MethodHandleResolver.ffiCallLinkCallerMethod()
+	 * to obtain a MemberName object plus appendix in a two element object array.
+	 */
+	VM_OutOfLineINL_Helpers::buildInternalNativeStackFrame(currentThread, method);
+	VM_VMHelpers::pushObjectInSpecialFrame(currentThread, nativeMH);
+	invokeCache = resolveFfiCallInvokeHandle(currentThread, nativeMH);
+	VM_VMHelpers::popObjectInSpecialFrame(currentThread);
+	if (VM_VMHelpers::exceptionPending(currentThread)) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		goto done;
+	}
+
+	VM_AtomicSupport::writeBarrier();
+	J9VMJAVALANGINVOKENATIVEMETHODHANDLE_SET_INVOKECACHE(currentThread, nativeMH, invokeCache);
+	VM_OutOfLineINL_Helpers::restoreInternalNativeStackFrame(currentThread);
+
+done:
+	VM_OutOfLineINL_Helpers::returnVoid(currentThread, 2);
```

Comment:
Now that the method is `static`, I think the `slotCount` should be 1.

Reply:
Agreed and fixed.